streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1932,6 +1932,35 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
   final def tap[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, _]): ZStream[R1, E1, A] =
     ZStream[R1, E1, A](self.process.map(_.tap(f(_).mapError(Some(_)))))
 
+  /**
+   * Groups elements into chunks of the specified size.
+   */
+  final def chunked(chunkSize: Int): ZStream[R, E, Chunk[A]] =
+    ZStream {
+      self.process.mapM { as =>
+        Ref.make[Option[Option[E]]](None).flatMap { stateRef =>
+          def loop(chunk: Chunk[A]): Pull[R, E, Chunk[A]] =
+            as.foldM(
+              success = { a =>
+                val chunk1 = chunk ++ Chunk.single(a)
```

Comment:
This would result in a left-associated "tree" of concats. Not the best structure for a chunk! Since the purpose of this implementation is to be performant, we can alternatively:
- preallocate an array when entering `loop`, set elements on it when pulling upstream and use `Chunk.fromArray` / `Chunk.fromArray(arr).slice(...)`
- Use `Pull.emit(chunk1.materialize)` which would copy the elements of the concatenated chunks into an array.

Reply:
The first option, although it is the fastest, would require a `ClassTag[A]`, so I chose the latter. Using a `List` would also work, not sure which one is slower, `Chunk.fromIterable(List.reverse)` or `Chunk.materialize`.